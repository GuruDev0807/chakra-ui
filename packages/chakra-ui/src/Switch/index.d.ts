import { BoxProps } from "../Box";
import * as React from "react";
import { Omit } from "../common-types";
import { VariantColor } from "../theme";

export interface ISwitch {
  /**
   * The size of the switch
   */
  size?: "sm" | "md" | "lg";
  /**
   * The background color of the switch when checked
   */
  color?: VariantColor;
  /**
   * The input name of the switch when used in a form
   */
  name?: string;
  /**
   * The value of the switch.
   */
  value?: string | number | boolean;
  /**
   * The children of the switch.
   */
  children?: React.ReactNode;
  /**
   * The aria-label of the switch for accessibility.
   */
  "aria-label"?: string;
  /**
   * The aria-labelledby of the switch for accessibility.
   */
  "aria-labelledby"?: string;
  /**
   * If `true`, set the switch to the checked state.
   */
  isChecked?: boolean;
  /**
   * If `true`, the switch will be initially checked.
   */
  defaultIsChecked?: boolean;
  /**
   * If `true`, set the disabled to the invalid state.
   */
  isDisabled?: boolean;
  /**
   * If `true`, set the switch to the invalid state.
   */
  isInvalid?: boolean;
}

export type SwitchProps = ISwitch & Omit<BoxProps, "defaultChecked" | "size">;

declare const Switch: React.FC<SwitchProps>;

export default Switch;
