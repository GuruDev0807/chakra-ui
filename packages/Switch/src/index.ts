export * from "./Switch";
export * from  "./Switch.hook";