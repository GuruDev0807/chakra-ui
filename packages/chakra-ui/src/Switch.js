/** @jsx jsx */
import { jsx } from "@emotion/core";
import styled from "@emotion/styled";
import { themeGet } from "@styled-system/theme-get";
import { forwardRef } from "react";
import { Box } from "./Layout";
import VisuallyHidden from "./VisuallyHidden";
import { useUIMode } from "./ThemeProvider";

const Knob = styled(Box)`
  background: white;
  transition: transform 250ms;
  transform: translateX(0);
`;

const knobTranslateX = props => {
  let { width, height } = props.theme.sizes.switch[props.switchSize];
  return `transform: translateX(calc(${width} - ${height}))`;
};

const Track = styled(Box)`
  display: inline-flex;
  cursor: pointer;
  flex-shrink: 0;
  background-color: ${props =>
    props.mode === "dark"
      ? themeGet("colors.alpha.400")(props)
      : themeGet("colors.gray.300")(props)};
  transition: all 0.2s;
  box-sizing: content-box;
  padding: 2px;

  input[type="checkbox"]:checked + & {
    background-color: ${props =>
      themeGet(`colors.${props.switchColor}.500`)(props)};
    ${Knob} {
      ${knobTranslateX}
    }
  }

  input[type="checkbox"]:disabled + & {
    opacity: 0.5;
  }

  input[type="checkbox"]:focus + & {
    box-shadow: ${themeGet("shadows.focusring")};
  }
`;

const Switch = forwardRef(
  (
    {
      id,
      isChecked,
      size,
      defaultChecked,
      isDisabled,
      children,
      color,
      ...rest
    },
    ref
  ) => {
    const {mode} = useUIMode();
    return (
      <Box as="label" display="inline-block" verticalAlign="middle">
        <VisuallyHidden
          as="input"
          id={id}
          ref={ref}
          checked={isChecked}
          defaultChecked={defaultChecked}
          type="checkbox"
          disabled={isDisabled}
          {...rest}
        />
        <Track
          borderRadius="round"
          switchColor={color}
          switchSize={size}
          mode={mode}
          width={`switch.${size}.width`}
          height={`switch.${size}.height`}
        >
          <Knob borderRadius="round" size={`switch.${size}.height`} />
        </Track>
      </Box>
    );
  }
);

Switch.defaultProps = {
  color: "blue",
  size: "md"
};

export default Switch;
