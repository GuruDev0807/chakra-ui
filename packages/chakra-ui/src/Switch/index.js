/** @jsx jsx */
import { jsx } from "@emotion/core";
import { forwardRef } from "react";
import Box from "../Box";
import { useColorMode } from "../ColorModeProvider";
import ControlBox from "../ControlBox";
import VisuallyHidden from "../VisuallyHidden";

const switchSizes = {
  sm: {
    width: "1.375rem",
    height: "0.75rem",
  },
  md: {
    width: "1.875rem",
    height: "1rem",
  },
  lg: {
    width: "2.875rem",
    height: "1.5rem",
  },
};

const Switch = forwardRef(
  (
    {
      id,
      name,
      value,
      "aria-label": ariaLabel,
      "aria-labelledby": ariaLabelledBy,
      color,
      defaultIsChecked,
      isChecked,
      size,
      isDisabled,
      isInvalid,
      onChange,
      onBlur,
      onFocus,
      children,
      ...rest
    },
    ref,
  ) => {
    const { colorMode } = useColorMode();
    const width = switchSizes[size] && switchSizes[size]["width"];
    const height = switchSizes[size] && switchSizes[size]["height"];

    const stylesProps = {
      rounded: "full",
      justifyContent: "flex-start",
      width,
      height,
      bg: colorMode === "dark" ? "whiteAlpha.400" : "gray.300",
      boxSizing: "content-box",
      p: "2px",
      _checked: {
        bg: `${color}.500`,
      },
      _child: {
        transform: `translateX(0)`,
      },
      _checkedAndChild: {
        transform: `translateX(calc(${width} - ${height}))`,
      },
      _focus: {
        boxShadow: "outline",
      },
      _hover: {
        cursor: "pointer",
      },
      _checkedAndHover: {
        cursor: "pointer",
      },
      _disabled: {
        opacity: 0.4,
        cursor: "not-allowed",
      },
    };

    return (
      <Box as="label" display="inline-block" verticalAlign="middle" {...rest}>
        <VisuallyHidden
          as="input"
          type="checkbox"
          aria-label={ariaLabel}
          aria-labelledby={ariaLabelledBy}
          id={id}
          ref={ref}
          name={name}
          value={value}
          aria-invalid={isInvalid}
          defaultChecked={defaultIsChecked}
          onChange={onChange}
          onBlur={onBlur}
          onFocus={onFocus}
          checked={isChecked}
          disabled={isDisabled}
        />
        <ControlBox {...stylesProps}>
          <Box
            bg="white"
            transition="transform 250ms"
            rounded="full"
            size={height}
          />
        </ControlBox>
      </Box>
    );
  },
);

Switch.displayName = "Switch";

Switch.defaultProps = {
  color: "blue",
  size: "md",
};

export default Switch;
