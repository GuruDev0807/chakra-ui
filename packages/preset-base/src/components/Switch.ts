import { ComponentTheme, getModeColor } from "./utils"

const sizes: ComponentTheme["sizes"] = {
  sm: {
    Track: {
      width: "1.375rem",
      height: "0.75rem",
    },
    Thumb: {
      width: "0.75rem",
      height: "0.75rem",
      _checked: { transform: `translateX(0.625rem)` },
    },
  },
  md: {
    Track: {
      width: "1.875rem",
      height: "1rem",
    },
    Thumb: {
      width: "1rem",
      height: "1rem",
      _checked: { transform: `translateX(0.875rem)` },
    },
  },
  lg: {
    Track: {
      width: "2.875rem",
      height: "1.5rem",
    },
    Thumb: {
      width: "1.5rem",
      height: "1.5rem",
      _checked: { transform: `translateX(1.375rem)` },
    },
  },
}

const Switch: ComponentTheme = {
  defaultProps: {
    size: "md",
  },
  baseStyle: (props: any) => ({
    Track: {
      borderRadius: "full",
      justifyContent: "flex-start",
      boxSizing: "content-box",
      padding: "2px",
      cursor: "pointer",
      bg: getModeColor(props, "gray.300", "whiteAlpha.400"),
      _focus: { boxShadow: "outline" },
      _disabled: {
        opacity: 0.4,
        cursor: "not-allowed",
      },
      _checked: { bg: `${props.variantColor}.500` },
    },
    Thumb: {
      bg: "white",
      transition: "transform 250ms",
      borderRadius: "full",
      transform: `translateX(0)`,
    },
  }),
  sizes,
}

export default Switch
