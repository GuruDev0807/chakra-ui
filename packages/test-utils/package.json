{
  "name": "@chakra-ui/test-utils",
  "version": "1.0.0-next.6",
  "private": true,
  "description": "Testing utilities for Chakra UI's components",
  "keywords": [
    "chakra",
    "test utils",
    "testing",
    "utils",
    "testing library",
    "react testing library"
  ],
  "author": "Segun Adebayo <sage@adebayosegun.com>",
  "homepage": "https://github.com/chakra-ui/chakra-ui#readme",
  "license": "MIT",
  "main": "src/index.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/chakra-ui/chakra-ui.git"
  },
  "bugs": {
    "url": "https://github.com/chakra-ui/chakra-ui/issues"
  },
  "dependencies": {
    "@chakra-ui/system": "^1.0.0-next.6",
    "@chakra-ui/theme": "^1.0.0-next.6",
    "@testing-library/jest-dom": "5.1.1",
    "@testing-library/react": "10.4.5",
    "@testing-library/react-hooks": "3.2.1",
    "@testing-library/user-event": "12.0.11",
    "@types/jest-axe": "3.2.1",
    "jest-axe": "3.4.0",
    "react-test-renderer": "16.12.0"
  },
  "peerDependencies": {
    "react": "16.x"
  }
}
