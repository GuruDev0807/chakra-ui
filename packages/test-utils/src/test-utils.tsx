import { ChakraProvider } from "@chakra-ui/core"
import "@testing-library/jest-dom/extend-expect"
import {
  render as RTLRender,
  RenderOptions,
  fireEvent,
  RenderResult,
  Queries,
} from "@testing-library/react"
import * as React from "react"
import { toHaveNoViolations } from "jest-axe"
import serializer from "jest-emotion"

expect.addSnapshotSerializer(serializer)
expect.extend(toHaveNoViolations)

const AllProviders: React.FC = ({ children }) => (
  <ChakraProvider resetCSS>{children}</ChakraProvider>
)

export function render(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">,
): RenderResult

export function render(ui: React.ReactElement, options?: RenderOptions) {
  return RTLRender(ui, { wrapper: AllProviders, ...options })
}

export * from "@testing-library/react"

export {
  act as invoke,
  renderHook,
  RenderHookOptions,
  RenderHookResult,
} from "@testing-library/react-hooks"

export { default as userEvent } from "@testing-library/user-event"

export * from "jest-axe"

export const escape = (ui: HTMLElement) =>
  fireEvent.keyDown(ui, { key: "Escape", keyCode: 27 })
