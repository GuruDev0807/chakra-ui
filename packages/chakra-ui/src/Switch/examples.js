import { storiesOf } from "@storybook/react";
import React from "react";
import Box from "../Box";
import Stack from "../Stack";
import Switch from "../Switch";

const stories = storiesOf("Switch", module);

stories.addDecorator(story => {
  return (
    <Box maxWidth="lg" mx="auto" mt={6} p={6}>
      {story()}
    </Box>
  );
});

stories.add("Default", () => (
  <Stack inline>
    <Switch
      size="sm"
      onChange={e => console.log(e.target.checked)}
      color="green"
    />
    <Switch
      size="md"
      onChange={e => console.log(e.target.checked)}
      color="blue"
    />
    <Switch
      size="lg"
      isDisabled
      onChange={e => console.log(e.target.checked)}
      color="cyan"
    />
  </Stack>
));
