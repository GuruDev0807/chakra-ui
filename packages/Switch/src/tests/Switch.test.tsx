import React from "react"
import { userEvent, render } from "@chakra-ui/test-utils"
import Switch from "../Switch"

test("Switch renders correctly", () => {
  const utils = render(<Switch />)
  expect(utils.asFragment()).toMatchSnapshot()
})

test("", () => {})
