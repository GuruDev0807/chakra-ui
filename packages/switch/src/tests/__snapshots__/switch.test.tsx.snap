// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders correctly 1`] = `
<DocumentFragment>
  <label
    class="chakra-switch css-v7ibj3"
  >
    <input
      class="chakra-switch__input"
      style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
      type="checkbox"
      value=""
    />
    <div
      aria-hidden="true"
      class="chakra-switch__track css-1qiw92k"
    >
      <div
        class="chakra-switch__thumb css-6y46ei"
      />
    </div>
  </label>
</DocumentFragment>
`;
